File: packages/backend/spi-simulator/src/models/transaction.ts
export type TransactionStatus =
  | 'ACCEPTED'
  | 'REJECTED'
  | 'SETTLED'
  | 'RETURNED'

export interface Transaction {
  id: string
  endToEndId: string
  txId?: string
  amount: number
  creditorIspb: string
  creditorKey?: string
  creditorName?: string
  debtorIspb: string
  debtorKey?: string
  debtorName?: string
  status: TransactionStatus
  returnReason?: string
  returnRejectionReason?: string
  createdAt: string
  settledAt?: string
  returnedAt?: string
  originalEndToEndId?: string
}

const transactions: Map<string, Transaction> = new Map()

export function addTransaction(tx: Transaction): void {
  transactions.set(tx.endToEndId, tx)
}

export function getTransaction(endToEndId: string): Transaction | undefined {
  return transactions.get(endToEndId)
}

export function updateTransaction(
  endToEndId: string,
  updates: Partial<Transaction>
): Transaction | undefined {
  const tx = transactions.get(endToEndId)
  if (!tx) return undefined
  Object.assign(tx, updates)
  return tx
}

export function getAllTransactions(): Transaction[] {
  return Array.from(transactions.values())
}

export function clearTransactions(): void {
  transactions.clear()
}

File: packages/backend/spi-simulator-go/internal/models/transaction.go
package models

import "time"

type TransactionStatus string

const (
	StatusAccepted TransactionStatus = "ACCEPTED"
	StatusRejected TransactionStatus = "REJECTED"
	StatusSettled  TransactionStatus = "SETTLED"
	StatusReturned TransactionStatus = "RETURNED"
)

type Transaction struct {
	ID                   string            `json:"id"`
	EndToEndID           string            `json:"endToEndId"`
	TxID                 string            `json:"txId,omitempty"`
	Amount               float64           `json:"amount"`
	CreditorISPB         string            `json:"creditorIspb"`
	CreditorKey          string            `json:"creditorKey,omitempty"`
	CreditorName         string            `json:"creditorName"`
	DebtorISPB           string            `json:"debtorIspb"`
	DebtorKey            string            `json:"debtorKey,omitempty"`
	DebtorName           string            `json:"debtorName"`
	Status               TransactionStatus `json:"status"`
	ReturnReason         string            `json:"returnReason,omitempty"`
	ReturnRejectionReason string           `json:"returnRejectionReason,omitempty"`
	CreatedAt            time.Time         `json:"createdAt"`
	SettledAt            *time.Time        `json:"settledAt,omitempty"`
	ReturnedAt           *time.Time        `json:"returnedAt,omitempty"`
	OriginalEndToEndID   string            `json:"originalEndToEndId,omitempty"`
}
